server/controller/src/main/java/ai/starwhale/mlops/datastore/wal/WalLocalFileManager.java

```diff
@@ -0,0 +1,236 @@
+                            entry.getTableSchema(),
+                            entry.getRecordsCount());
+                }
+                if (this.currentFileSize + sizeTotal + entrySize > this.walMaxFileSize) {
+                    if (i == 0) {
+                        ++this.logFileIndex;
+                        this.initLogFile();
+                        return this.write(entries);
+                    }
+                    ret = i;
+                    break;
+                }
+                entry.writeDelimitedTo(this.outputStream);
+                sizeTotal += entrySize;
+            }
+            this.outputStream.flush();
+        } catch (IOException e) {
+            if (this.outputStream != null) {
+                try {
+                    this.outputStream.close();
+                } catch (IOException e1) {
+                    log.error("failed to close", e1);
+                }
+                this.outputStream = null;
+            }
+            throw e;
+        }
+        this.currentFileSize += sizeTotal;
+        var info = this.logFileMap.get(this.logFileIndex);
+        info.maxEntryId = entries.get(ret - 1).getId();
+        info.size += sizeTotal;
+        return ret;
+    }
+
+    public Iterator<WalEntry> readAll(int maxUncompressedSize) {
+        return new WalEntryReader(this.indexes, maxUncompressedSize, true) {
+
+            @Override
+            public WalEntry next() {
+                var entry = super.next();
+                var info = WalLocalFileManager.this.logFileMap.computeIfAbsent(this.getCurrentLogIndex(),
+                        LogFileInfo::new);
+                info.maxEntryId = entry.getId();
+                info.size += CodedOutputStream.computeMessageSizeNoTag(entry);
+                if (this.getCurrentLogIndex() == WalLocalFileManager.this.logFileIndex) {
+                    WalLocalFileManager.this.currentFileSize = info.size;
+                }
+                return entry;
+            }
+
+            protected byte[] readLogFile(long logIndex) {
+                var path = WalLocalFileManager.this.getLocalFilePath(logIndex);
+                try {
+                    return Files.readAllBytes(path);
+                } catch (IOException e) {
+                    throw new SwProcessException(ErrorType.DATASTORE, "fail to read " + path, e);
+                }
+            }
+        };
+    }
+
+    private Path getLocalFilePath(long index) {
+        return this.cacheDir.resolve("wal.log." + index);
+    }
+
+    private void initLogFile() {
+        if (this.outputStream != null) {
+            try {
+                this.outputStream.close();
+            } catch (IOException e) {
+                log.error("failed to close", e);
+            }
+        }
+        this.currentFileSize = WalManager.HEADER.length;
+        var info = new LogFileInfo(this.logFileIndex);
+        this.logFileMap.put(this.logFileIndex, info);
```

Comment:
The `logFileMap` will increase without limit

Reply:
Done